globus_automate_client/models.py

```diff
@@ -0,0 +1,389 @@
+        if len(set_fields) > 1:
+            raise ValueError(
+                f"Only one of these fields may be set: {set_fields.keys()}"
+            )
+        if len(set_fields) == 0:
+            raise ValueError(f"At least one field must be set: {fields}")
+
+        return values
+
+
+FlowState = te.Annotated[
+    t.Union[
+        PassState, WaitState, ActionState, FailState, ChoiceState, ExpressionEvalState
+    ],
+    Field(discriminator="Type"),
+]
+
+
+class FlowDefinition(BaseModel):
+    StartAt: str
+    States: t.Dict[str, FlowState]
+    Comment: t.Optional[str]
+
+    @validator("States")
+    def validate_states(cls, v):
+        if len(v) < 1:
+            raise ValueError('At least one "State" is required')
+
+        invalid_state_names = [
+            state_name
+            for state_name in v.keys()
+            if len(state_name) > 128 or len(state_name) == 0
+        ]
+        if invalid_state_names:
+            raise ValueError(
+                f"The following states have invalid names: {', '.join(invalid_state_names)}"
+            )
+        return v
+
+    @root_validator(skip_on_failure=True)
+    def validate_start_at_is_a_state(cls, values):
+        states, start_at = values.get("States"), values.get("StartAt")
+        if states and start_at not in states:
+            raise ValueError(f'StartAt state "{start_at}" is not defined in States')
+        return values
+
+    @root_validator(skip_on_failure=True)
+    def validate_all_transtion_states_exist(cls, values):
+        states = values.get("States", {})
+        defined_states = set(states.keys())
+        referenced_states = {values["StartAt"]}
+
+        for _, state in states.items():
+            if state.Type not in {StateType.Choice, StateType.Fail}:
+                # Ensure that all state's with a Next property have a valid transition
+                next_state = state.Next
+                if next_state is not None:
+                    referenced_states.add(next_state)
+
+            # Ensure that the Catch conditions in an Action state refer to a
+            # valid transition
+            if state.Type == StateType.Action:
+                if state.Catch:
+                    for catcher in state.Catch:
+                        referenced_states.add(catcher.Next)
+
+            # Ensure that the choice state rules have valid transitions
+            elif state.Type == StateType.Choice:
+                if state.Default:
+                    referenced_states.add(state.Default)
+                for choice in state.Choices:
+                    referenced_states.add(choice.Next)
+
+        undefined_states = referenced_states - defined_states
+        unreferenced_states = defined_states - referenced_states
+        if undefined_states:
+            raise ValueError(
+                f'Flow definition contained a reference to unknown state(s): {",".join(undefined_states)}'
```

Comment:
I think `{", ".join(undefined_states)}` will read better (note the space).

Reply:
Isn't that about what we'd get if we just did the equivalent of `str(undefined_states)`?

Reply:
Added!

Reply:
@jimPruyne for some reason i wasn't able to response directly to your comment about string-ing a list. Looks like python isn't as clever as it pretends to be:

```
>>> l = ["a", "bb", "ccc"]
>>> str(l)
"['a', 'bb', 'ccc']"
```